File: pystreams/src/pystreams/risk/enforce_handler.py
"""Inline Presidio enforcement scans with a correlated Redis reply.

Python counterpart of the gitleaks enforcement handler
(``server/internal/scanners/gitleaks/enforce_handler.go``), consuming the
dedicated ``gram.risk.v1.PresidioEnforcement`` topic. The handler scans one
request and writes an ``EnforcementReply`` to the replica inbox named by the
reply URN. Replies never carry raw matched text: findings hold masked previews
and tenant-peppered fingerprints only, and logs report entity types and counts.

Ack/nack policy mirrors the Go handler: stale requests, scan failures, and
reply-write failures are ACKed (redelivery cannot rescue an inline scan, and a
nack would eventually park raw request content in the DLQ), while undecodable
or structurally invalid requests raise so the restricted forensic DLQ retains
what could not be interpreted.
"""

from __future__ import annotations

import time
import uuid
from datetime import UTC, datetime
from typing import Final

import structlog
from gram.risk.v1 import enforcement_reply_pb2, presidio_enforcement_pb2
from gram_infra.pubsub.subscriber import MessageMetadata
from opentelemetry import metrics

from pystreams.risk import maskdisplay
from pystreams.risk.fingerprint import Fingerprinter, encode_fingerprint
from pystreams.risk.replywriter import ReplyWriter, parse_reply_urn
from pystreams.risk.scanner import (
    DEFAULT_SCORE_THRESHOLD,
    Detection,
    Scanner,
    ScanSlotTimeout,
)

# Maximum useful age of an inline scan request; the dispatcher's wait ceiling
# means anything older can no longer satisfy a caller.
DEFAULT_MAX_REQUEST_AGE_SECONDS: Final = 30.0

# Per-message content budget, matching replyinbox.MaxContentBytes on the Go
# dispatcher; a request that bypasses the dispatcher must not buy an unbounded
# scan.
MAX_CONTENT_BYTES: Final = 50 * 1024

# Transport-metadata key carrying the request's return address; must match
# requestreply.ReplyURNAttribute on the Go side.
REPLY_URN_ATTRIBUTE: Final = "gram-reply-urn"

_MAX_REPLY_REASON_CHARS: Final = 256

# Presidio pii.* rule ids mapped to their canonical category, mirroring the
# classification order in server/internal/risk/categories (first match wins;
# anything else with source "presidio" is pii).
_GOVERNMENT_ID_RULE_IDS: Final = frozenset(
    {
        "pii.us_ssn",
        "pii.us_passport",
        "pii.us_itin",
        "pii.uk_nhs",
        "pii.uk_nino",
        "pii.uk_passport",
        "pii.es_nif",
        "pii.it_fiscal_code",
        "pii.au_tfn",
        "pii.in_pan",
        "pii.in_aadhaar",
        "pii.sg_nric_fin",
    }
)
_HEALTHCARE_RULE_IDS: Final = frozenset(
    {
        "pii.medical_license",
        "pii.us_mbi",
        "pii.us_npi",
        "pii.medical_disease_disorder",
        "pii.medical_medication",
        "pii.medical_therapeutic_procedure",
        "pii.medical_clinical_event",
        "pii.medical_biological_attribute",
        "pii.medical_family_history",
    }
)
_OFF_POLICY_RULE_IDS: Final = frozenset(
    {
        "pii.harmful_content_request",
        "pii.policy_violation",
        "pii.unauthorized_action",
        "pii.topic_boundary_violation",
    }
)

_meter = metrics.get_meter("pystreams.risk.enforce")
_stale_dropped = _meter.create_counter(
    "risk.enforcement.presidio.stale_dropped",
    unit="{request}",
    description=(
        "Presidio enforcement requests acknowledged without scanning because "
        "they were stale"
    ),
)
_reply_write_errors = _meter.create_counter(
    "risk.enforcement.presidio.reply_write_errors",
    unit="{error}",
    description=("Presidio enforcement reply writes acknowledged after Redis failure"),
)


class MalformedEnforcementRequest(ValueError):
    """A structurally invalid request; raised so it lands in the forensic DLQ."""


class PresidioEnforceHandler:
    """Scans one inline request and writes a safe correlated reply."""

    def __init__(
        self,
        logger: structlog.stdlib.BoundLogger,
        writer: ReplyWriter,
        scanner: Scanner,
        fingerprinter: Fingerprinter,
        max_request_age_seconds: float = DEFAULT_MAX_REQUEST_AGE_SECONDS,
    ) -> None:
        self._logger = logger
        self._writer = writer
        self._scanner = scanner
        self._fingerprinter = fingerprinter
        if max_request_age_seconds <= 0:
            max_request_age_seconds = DEFAULT_MAX_REQUEST_AGE_SECONDS
        self._max_request_age = max_request_age_seconds
        self._consumer_id = str(uuid.uuid4())

    async def handle(
        self,
        message: presidio_enforcement_pb2.PresidioEnforcement,
        meta: MessageMetadata,
    ) -> None:
        try:
            created_at = datetime.fromisoformat(message.created_at)
            if created_at.tzinfo is None:
                raise ValueError("enforcement created_at must include a timezone")
        except ValueError as exc:
            raise MalformedEnforcementRequest("parse enforcement created_at") from exc
        age = (datetime.now(UTC) - created_at).total_seconds()
        # Symmetric window: a far-future stamp is as suspect as a stale one,
        # while ordinary clock skew stays within the allowance.
        if abs(age) > self._max_request_age:
            _stale_dropped.add(1)
            return
        if not message.organization_id:
            raise MalformedEnforcementRequest("enforcement organization id is required")
        if not message.project_id:
            raise MalformedEnforcementRequest("enforcement project id is required")
        # The reply address rides transport metadata, mirroring Go's
        # requestreply.ReplyURNAttribute; payloads carry no routing.
        reply_urn = meta.attributes.get(REPLY_URN_ATTRIBUTE, "")
        try:
            _, scan_id = parse_reply_urn(reply_urn)
        except ValueError as exc:
            raise MalformedEnforcementRequest("parse enforcement reply urn") from exc

        started = time.perf_counter()
        detections: list[Detection] = []
        status = enforcement_reply_pb2.ENFORCEMENT_STATUS_OK
        reason = ""
        if len(message.content.encode()) > MAX_CONTENT_BYTES:
            status = enforcement_reply_pb2.ENFORCEMENT_STATUS_ERROR
            reason = "enforcement content exceeds the maximum byte budget"
        elif not 0.0 <= message.score_threshold <= 1.0:
            # A negative threshold admits noise; above 1.0 suppresses every
            # finding. Neither is a scan worth running.
            status = enforcement_reply_pb2.ENFORCEMENT_STATUS_ERROR
            reason = "enforcement score threshold is outside 0.0-1.0"
        else:
            requested = list(message.entities) or None
            score_threshold = message.score_threshold or DEFAULT_SCORE_THRESHOLD
            try:
                detections = await self._scanner.scan(
                    message.content, requested, score_threshold
                )
            except ScanSlotTimeout:
                # Unlike the batch path, requeueing cannot rescue an inline
                # scan: reply ERROR now so the waiter applies its configured
                # failure mode instead of burning the deadline.
                status = enforcement_reply_pb2.ENFORCEMENT_STATUS_ERROR
                reason = "presidio scan slot timeout"
            except Exception as exc:
                # Only the exception type: an error string can echo the
                # scanned content, which never leaves this handler.
                status = enforcement_reply_pb2.ENFORCEMENT_STATUS_ERROR
                reason = "presidio scan failed: " + type(exc).__name__

        findings: list[enforcement_reply_pb2.EnforcementFinding] = []
        if status == enforcement_reply_pb2.ENFORCEMENT_STATUS_OK:
            try:
                findings = self._build_findings(message.organization_id, detections)
            except Exception as exc:
                status = enforcement_reply_pb2.ENFORCEMENT_STATUS_ERROR
                reason = "fingerprint enforcement finding"
                findings = []
                self._logger.error(
                    "fingerprint presidio enforcement finding",
                    request_id=message.request_id,
                    error_type=type(exc).__name__,
                )

        reply = enforcement_reply_pb2.EnforcementReply(
            correlation_id=scan_id,
            scanner=enforcement_reply_pb2.ENFORCEMENT_SCANNER_PRESIDIO,
            status=status,
            reason=reason[:_MAX_REPLY_REASON_CHARS],
            findings=findings,
            diagnostics=enforcement_reply_pb2.EnforcementDiagnostics(
                scan_duration_ms=int((time.perf_counter() - started) * 1000),
                consumer_id=self._consumer_id,
                delivery_attempt=meta.delivery_attempt or 0,
            ),
        )
        try:
            await self._writer.write(reply_urn, reply)
        except Exception as exc:
            _reply_write_errors.add(1)
            self._logger.error(
                "write presidio enforcement reply; acknowledging request",
                request_id=message.request_id,
                error_type=type(exc).__name__,
            )
            return

        await self._logger.adebug(
            "presidio enforcement scan complete",
            request_id=message.request_id,
            detections=len(detections),
            status=enforcement_reply_pb2.EnforcementStatus.Name(status),
        )

    def _build_findings(
        self, organization_id: str, detections: list[Detection]
    ) -> list[enforcement_reply_pb2.EnforcementFinding]:
        findings: list[enforcement_reply_pb2.EnforcementFinding] = []
        for d in detections:
            rule_id = "pii." + d.entity_type.lower()
            sum_, _ = self._fingerprinter.tenanted_hs256(
                organization_id, d.match.encode()
            )
            findings.append(
                enforcement_reply_pb2.EnforcementFinding(
                    rule_id=rule_id,
                    category=_classify(rule_id),
                    score=d.confidence,
                    start_pos=d.start_pos,
                    end_pos=d.end_pos,
                    surface="content",
                    masked_preview=maskdisplay.display(rule_id, d.match),
                    fingerprint=encode_fingerprint(sum_),
                )
            )
        return findings


def _classify(rule_id: str) -> str:
    if rule_id in maskdisplay.FINANCIAL_RULE_IDS:
        return "financial"
    if rule_id in _GOVERNMENT_ID_RULE_IDS:
        return "government_ids"
    if rule_id in _HEALTHCARE_RULE_IDS:
        return "healthcare"
    if rule_id in _OFF_POLICY_RULE_IDS:
        return "off_policy"
    return "pii"

File: server/internal/scanners/gitleaks/enforce_handler.go
package gitleaks

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"time"

	"github.com/google/uuid"
	"go.opentelemetry.io/otel/metric"

	riskv1 "github.com/speakeasy-api/gram/infra/gen/gram/risk/v1"
	"github.com/speakeasy-api/gram/infra/pkg/gcp"
	"github.com/speakeasy-api/gram/server/internal/attr"
	"github.com/speakeasy-api/gram/server/internal/conv"
	"github.com/speakeasy-api/gram/server/internal/requestreply"
	"github.com/speakeasy-api/gram/server/internal/risk/categories"
	"github.com/speakeasy-api/gram/server/internal/risk/enforcereply"
	"github.com/speakeasy-api/gram/server/internal/risk/maskdisplay"
	"github.com/speakeasy-api/gram/server/internal/scanners"
)

const (
	// DefaultMaxRequestAge is the maximum useful age of an inline scan request.
	DefaultMaxRequestAge = 30 * time.Second
	maxReplyReasonRunes  = 256
)

// FingerprintFinding returns a canonically encoded, tenant-scoped fingerprint.
// Production callers must adapt risk.Fingerprinter.TenantedHS256 and
// risk.EncodeFingerprint here; the callback avoids a package import cycle.
type FingerprintFinding func(tenantID string, message []byte) (string, error)

// EnforceHandlerConfig controls inline request freshness.
type EnforceHandlerConfig struct {
	// MaxRequestAge drops requests that can no longer satisfy an inline caller.
	MaxRequestAge time.Duration
}

// EnforceHandler scans one inline request and writes a safe correlated reply.
type EnforceHandler struct {
	logger        *slog.Logger
	writer        requestreply.ReplyBroker[*riskv1.EnforcementReply]
	scanner       *Scanner
	fingerprint   FingerprintFinding
	metrics       enforceHandlerMetrics
	consumerID    string
	maxRequestAge time.Duration
}

// NewEnforceHandler builds the gitleaks enforcement subscription handler.
func NewEnforceHandler(
	logger *slog.Logger,
	meterProvider metric.MeterProvider,
	writer requestreply.ReplyBroker[*riskv1.EnforcementReply],
	fingerprint FingerprintFinding,
	cfg EnforceHandlerConfig,
) (*EnforceHandler, error) {
	if writer == nil {
		return nil, errors.New("gitleaks enforcement reply writer is required")
	}
	if fingerprint == nil {
		return nil, errors.New("gitleaks enforcement fingerprint function is required")
	}
	if cfg.MaxRequestAge <= 0 {
		cfg.MaxRequestAge = DefaultMaxRequestAge
	}
	metrics, err := newEnforceHandlerMetrics(meterProvider)
	if err != nil {
		return nil, err
	}
	return &EnforceHandler{
		logger:        logger.With(attr.SlogComponent("gitleaks-enforcer")),
		writer:        writer,
		scanner:       NewScanner(),
		fingerprint:   fingerprint,
		metrics:       metrics,
		consumerID:    uuid.NewString(),
		maxRequestAge: cfg.MaxRequestAge,
	}, nil
}

// Handle ACKs completed and stale scans. Malformed requests return an error so
// the restricted forensic DLQ retains evidence that could not be interpreted.
func (h *EnforceHandler) Handle(ctx context.Context, m *riskv1.GitleaksEnforcement, meta gcp.MessageMetadata) error {
	createdAt, err := time.Parse(time.RFC3339Nano, m.GetCreatedAt())
	if err != nil {
		return fmt.Errorf("parse enforcement created_at: %w", err)
	}
	if time.Since(createdAt) > h.maxRequestAge {
		h.metrics.staleDropped.Add(ctx, 1)
		return nil
	}
	if m.GetOrganizationId() == "" {
		return errors.New("enforcement organization id is required")
	}
	if m.GetProjectId() == "" {
		return errors.New("enforcement project id is required")
	}
	replyURN := meta.Attributes[requestreply.ReplyURNAttribute]
	if replyURN == "" {
		return errors.New("enforcement reply urn attribute is required")
	}
	_, correlationID, err := enforcereply.ParseReplyURN(replyURN)
	if err != nil {
		return fmt.Errorf("parse enforcement reply urn: %w", err)
	}

	started := time.Now()
	var findings []scanners.Finding
	status := riskv1.EnforcementStatus_ENFORCEMENT_STATUS_OK
	reason := ""
	if len(m.GetContent()) > enforcereply.MaxContentBytes {
		// Dispatcher enforces this budget too; a request that bypasses it must
		// not buy an unbounded scan.
		status = riskv1.EnforcementStatus_ENFORCEMENT_STATUS_ERROR
		reason = fmt.Sprintf("enforcement content is %d bytes; maximum is %d bytes", len(m.GetContent()), enforcereply.MaxContentBytes)
	} else {
		var scanErr error
		findings, scanErr = h.scanner.Scan(ctx, m.GetContent())
		if scanErr != nil {
			status = riskv1.EnforcementStatus_ENFORCEMENT_STATUS_ERROR
			reason = scanErr.Error()
		}
	}

	replyFindings := make([]*riskv1.EnforcementFinding, 0, len(findings))
	for _, finding := range findings {
		fingerprint, fingerprintErr := h.fingerprint(m.GetOrganizationId(), []byte(finding.Match))
		if fingerprintErr != nil {
			h.logger.ErrorContext(ctx, "fingerprint gitleaks enforcement finding", attr.SlogError(fingerprintErr))
			status = riskv1.EnforcementStatus_ENFORCEMENT_STATUS_ERROR
			reason = "fingerprint enforcement finding"
			replyFindings = nil
			break
		}
		replyFindings = append(replyFindings, riskv1.EnforcementFinding_builder{
			RuleId:        new(finding.RuleID),
			Category:      new(string(categories.Classify(finding.Source, finding.RuleID))),
			Score:         new(finding.Confidence),
			StartPos:      new(conv.SafeInt32(finding.StartPos)),
			EndPos:        new(conv.SafeInt32(finding.EndPos)),
			Surface:       new(scanners.FindingSurface(finding.Source, finding.Field, finding.Path)),
			Field:         new(finding.Field),
			Path:          new(finding.Path),
			ToolCallId:    new(finding.McpLookupToolCallID),
			MaskedPreview: new(maskdisplay.Display(finding.Source, finding.RuleID, finding.Match)),
			Fingerprint:   new(fingerprint),
		}.Build())
	}

	deliveryAttempt := int32(0)
	if meta.DeliveryAttempt != nil {
		deliveryAttempt = conv.SafeInt32(*meta.DeliveryAttempt)
	}
	reply := riskv1.EnforcementReply_builder{
		CorrelationId: new(correlationID),
		Scanner:       new(riskv1.EnforcementScanner_ENFORCEMENT_SCANNER_GITLEAKS),
		Status:        new(status),
		Reason:        new(conv.TruncateString(reason, maxReplyReasonRunes)),
		Findings:      replyFindings,
		Diagnostics: riskv1.EnforcementDiagnostics_builder{
			ScanDurationMs:  new(time.Since(started).Milliseconds()),
			ConsumerId:      new(h.consumerID),
			DeliveryAttempt: new(deliveryAttempt),
		}.Build(),
	}.Build()
	if err := h.writer.Reply(ctx, replyURN, reply); err != nil {
		h.metrics.replyWriteErrors.Add(ctx, 1)
		h.logger.ErrorContext(ctx, "write gitleaks enforcement reply; acknowledging request", attr.SlogError(err))
		return nil
	}

	h.logger.DebugContext(ctx, "gitleaks enforcement scan complete", attr.SlogValueAny(map[string]any{
		"request_id": m.GetRequestId(),
		"detections": len(findings),
		"status":     status.String(),
	}))
	return nil
}
